1. True/False. Two reads with the same sample index, cellular barcode and UMI should be counted as a single read.

1. True/False. Quality control steps during the analysis of scRNA-seq data are crucial for the identification of accurate cell type clusters.

1. True/False. The library preparation method will determine which reads contain the sample index, cellular barcode, and UMI information.


1. Choose the most appropriate analysis. You do not expect a lot of cell type heterogeneity in your samples and you would like to determine the effects of a treatment on gene expression.
    - i. Bulk RNA-seq
    - i. scRNA-seq
    
1. Choose the most appropriate analysis. You would like to characterize the cell types present in a tissue.
    - i. Bulk RNA-seq
    - i. scRNA-seq
    
1. Choose the most appropriate analysis. You would like to determine the genes turned on during differentiation of a particular cell type.
    - i. Bulk RNA-seq
    - i. scRNA-seq
    
1. Choose the most appropriate analysis. You would like to determine the genes that are DE between two conditions and expect a lot of cell type heterogeneity; however, you can only afford a single sample per condition if you do scRNA-seq.
    - i. Bulk RNA-seq
    - i. scRNA-seq
    
1. Choose the most appropriate analysis. You would like to identify a strong biomarker for a disease.
    - i. Bulk RNA-seq
    - i. scRNA-seq
    
1. Choose the most appropriate analysis. You FACS sort your samples for a cell type of interest and would like to determine the DE of gene isoforms.
    - i. Bulk RNA-seq
    - i. scRNA-seq
